server/consumer.go

```diff
@@ -6045,14 +6046,19 @@ func (o *consumer) checkStateForInterestStream(ss *StreamState) error {
 		}
 		// Only ack though if no error and seq <= ack floor.
 		if err == nil && seq <= asflr {
-			mset.ackMsg(o, seq)
+			shouldRemove := mset.ackMsg(o, seq)
+			// Removing the message could fail. Lowering the floor here allows us to retry later if the removal failed.
+			if shouldRemove && retryAsflr == asflr {
```

Comment:
The comment does not match the logic here from what I can tell.

Reply:
Have rewritten it to make it clearer. Having a defaulted `var retryAsflr uint64` that only gets overwritten if it was its zero value (so only once/the first time).
```go
didRemove := mset.ackMsg(o, seq)
// Removing the message could fail.
// Overwrite retry floor (only the first time) to allow us to check next time if the removal was successful.
if didRemove && retryAsflr == 0 {
	retryAsflr = seq
}
```
And then overwriting from the zero value if all is good below the loop.
```go
// If retry floor was not overwritten, set to ack floor, we don't need to account for any retries below it.
if retryAsflr == 0 {
	retryAsflr = asflr
}
```

Reply:
If it does not get removed, say because of interest stream (meaning multiple consumers) because of the other consumer, we still can move ahead no?

Reply:
> If it does not get removed, say because of interest stream (meaning multiple consumers) because of the other consumer, we still can move ahead no?

I'm interpreting this in two different ways, so let me check both:

- If there are multiple consumers `didRemove=false` and we can move ahead. Only the last consumer that calls `ackMsg` will make `didRemove=true`, since all other consumers before it must have acked it.
- `retryAsflr = asflr` could actually be `retryAsflr = asflr + 1`, that prevents us retrying on `asflr` if we're not meant to remove the message. Have pushed a fix for that.

Reply:
I think we can always move ahead TBH..

Reply:
As discussed, I've turned this into draft for a little bit. I'll move the retry floor bit to https://github.com/nats-io/nats-server/pull/6140, then I'll ping you once this is ready to be merged again.